pandas/core/groupby/ops.py

```diff
@@ -198,6 +198,9 @@ def apply(self, f, data, axis=0):
             f_name not in base.plotting_methods
             and hasattr(splitter, "fast_apply")
             and axis == 0
+            # with MultiIndex, apply_frame_axis0 would raise InvalidApply
+            # TODO: can we make this check prettier?
+            and not splitter._get_sorted_data().index._has_complex_internals
```

Comment:
Do you generally know of a reason why we have `_has_complex_internals` as a Index property? I think only changed for a MultiIndex and only applicable in groupby space.

Might be easier and clearer to just do a isinstance(..., MultiIndex)

Reply:
I'm not aware of the history there no.

Reply:
Cool; just asking - probably a historical relic.

Haven't gone too deep on this one yet but will file comments (most likely) tomorrow

Reply:
_has_complex_internals predates me :->